pkg/validate/openshiftcluster_validatedynamic.go

```diff
@@ -112,51 +120,67 @@ func (dv *openShiftClusterDynamicValidator) Dynamic(ctx context.Context) error {
 		})
 	}
 
-	var pdpClient remotepdp.RemotePDPClient
-	spp := dv.oc.Properties.ServicePrincipalProfile
-
 	tenantID := dv.subscriptionDoc.Subscription.Properties.TenantID
-	options := dv.env.Environment().ClientSecretCredentialOptions()
-	spClientCred, err := azidentity.NewClientSecretCredential(
-		tenantID, spp.ClientID, string(spp.ClientSecret), options)
-	if err != nil {
-		return err
-	}
 	fpClientCred, err := dv.env.FPNewClientCertificateCredential(tenantID)
 	if err != nil {
 		return err
 	}
 
 	aroEnv := dv.env.Environment()
-	pdpClient = remotepdp.NewRemotePDPClient(
+	pdpClient := remotepdp.NewRemotePDPClient(
 		fmt.Sprintf(aroEnv.Endpoint, dv.env.Location()),
 		aroEnv.OAuthScope,
 		fpClientCred,
 	)
 
 	scopes := []string{dv.env.Environment().ResourceManagerScope}
-	err = ensureAccessTokenClaims(ctx, spClientCred, scopes)
-	if err != nil {
-		return err
-	}
-	spAuthorizer := azidext.NewTokenCredentialAdapter(spClientCred, scopes)
-
-	spDynamic := dynamic.NewValidator(
-		dv.log,
-		dv.env,
-		dv.env.Environment(),
-		dv.subscriptionDoc.ID,
-		spAuthorizer,
-		spp.ClientID,
-		dynamic.AuthorizerClusterServicePrincipal,
-		spClientCred,
-		pdpClient,
-	)
+	var spDynamic dynamic.Dynamic
+	if dv.oc.Properties.PlatformWorkloadIdentityProfile == nil || dv.oc.Properties.ServicePrincipalProfile != nil {
+		// SP validation
+		spp := dv.oc.Properties.ServicePrincipalProfile
+		options := dv.env.Environment().ClientSecretCredentialOptions()
+		spClientCred, err := azidentity.NewClientSecretCredential(
+			tenantID, spp.ClientID, string(spp.ClientSecret), options)
+		if err != nil {
+			return err
+		}
+		err = ensureAccessTokenClaims(ctx, spClientCred, scopes)
+		if err != nil {
+			return err
+		}
 
-	// SP validation
-	err = spDynamic.ValidateServicePrincipal(ctx, spClientCred)
-	if err != nil {
-		return err
+		spDynamic = dynamic.NewValidator(
+			dv.log,
+			dv.env,
+			dv.env.Environment(),
+			dv.subscriptionDoc.ID,
+			dv.fpAuthorizer,
+			&spp.ClientID,
+			dynamic.AuthorizerClusterServicePrincipal,
+			spClientCred,
+			pdpClient,
+		)
+		err = spDynamic.ValidateServicePrincipal(ctx, spClientCred)
+		if err != nil {
+			return err
+		}
+	} else {
+		// PlatformWorkloadIdentity and ClusterMSIIdentity Validation
+		spDynamic = dynamic.NewValidator(
+			dv.log,
+			dv.env,
+			dv.env.Environment(),
+			dv.subscriptionDoc.ID,
+			nil,
+			nil,
+			dynamic.AuthorizerClusterServicePrincipal,
```

Comment:
This authorizer type gets included in error messages, so if we use the existing `AuthorizerClusterServicePrincipal`, we're going to print some misleading stuff in error messages about a "service principal" that's missing permissions. I think it might be a good idea to define a new `AuthorizerType` for managed identities and pass it in here. Your thoughts?

Reply:
Added a new `AuthorizerType`.